### PR TITLE
#gh11588 prevent a HU from being added to the same return request multiple times

### PR DESCRIPTION
**Pull request by adi-stefan**
https://github.com/metasfresh/metasfresh/issues/11588

### CODE REVIEW

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inout/impl/HUInOutBL.java [hunk 1]
```diff
@@ -373,7 +387,38 @@ public void copyAssignmentsToReversal(@NonNull final org.compiere.model.I_M_InOu
...
+
+		final I_M_HU hu = handlingUnitsDAO.getById(huId);
+		final I_M_HU_Attribute serialNoAttr = huAttributesDAO.retrieveAttribute(hu, serialNoAttributeId);
+		if (serialNoAttr == null)
+		{
+			//no S/N defined. Should not be a valid scenario
+			return false;
+		}
+		return !queryBL.createQueryBuilder(I_M_HU_Assignment.class).addOnlyActiveRecordsFilter()
+				.addEqualsFilter(I_M_HU_Assignment.COLUMNNAME_AD_Table_ID, inOutDAO.getInOutTableId())
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inout/impl/HUInOutBL.java:411]: avoid using QueryBL outside repositories.

basically here, u shall do something like
* de.metas.handlingunits.inout.IHUInOutBL#getHUIdsByInOutLineIds - to retrieve the assigned HUIds. Consider introducing getHUIdsByInOutId
* use de.metas.handlingunits.impl.HandlingUnitsBL#getTopLevelHUs(java.util.Collection<de.metas.handlingunits.HuId>) - to get the top level HUIds
* having the top level HU Ids assigned to our InOut we can
* de.metas.handlingunits.impl.HandlingUnitsBL#createHUQueryBuilder
    * addOnlyHUIds(out top level HU Ids)
    * addOnlyWithAttribute(AttributeConstants.ATTR_SerialNo, "THE_SERIAL_NO")
* check if there is any match.

File: backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/SerialNoBL.java [hunk 2]
```diff
@@ -53,4 +55,13 @@
 				? Optional.of(serialNo)
 				: Optional.empty();
 	}
+
+	@Override
+	public Optional<AttributeId> getSerialNoAttributeId()
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/SerialNoBL.java:60]: not sure if this method will be needed after doing other suggested changes

File: backend/de.metas.business/src/main/java/de/metas/inout/impl/InOutDAO.java [hunk 3]
```diff
@@ -421,4 +424,10 @@ public void save(@NonNull final I_M_InOutLine inoutLine)
 	{
 		InterfaceWrapperHelper.saveRecord(inoutLine);
 	}
+
+	@Override
+	public AdTableId getInOutTableId()
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/inout/impl/InOutDAO.java:429]: not sure if this method will be needed after doing other suggested changes

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inout/impl/HUInOutBL.java (hunk 1 above)
Comment (adi-stefan) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inout/impl/HUInOutBL.java]: refactored as suggested.

File: backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/SerialNoBL.java (hunk 2 above)
Comment (adi-stefan) [reply backend/de.metas.business/src/main/java/org/adempiere/mm/attributes/api/impl/SerialNoBL.java:60]: still needed.

File: backend/de.metas.business/src/main/java/de/metas/inout/impl/InOutDAO.java (hunk 3 above)
Comment (adi-stefan) [reply backend/de.metas.business/src/main/java/de/metas/inout/impl/InOutDAO.java]: not needed, removed.

**Pull request merged**